security-csrf/src/main/java/io/micronaut/security/csrf/validator/RepositoryCsrfTokenValidator.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2017-2024 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.security.csrf.validator;
+
+import io.micronaut.context.annotation.Requires;
+import io.micronaut.security.csrf.generator.CsrfHmacTokenGenerator;
+import io.micronaut.security.csrf.repository.CsrfTokenRepository;
+import jakarta.inject.Singleton;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Optional;
+import java.security.MessageDigest;
+
+/**
+ * {@link CsrfTokenValidator} implementation that uses a {@link CsrfTokenRepository}.
+ * First attempts to retrieve a token from a {@link CsrfTokenRepository} and if found validates it against the supplied token.
+ * @param <T> Request
+ * @since 4.11.0
+ * @author Sergio del Amo
+ */
+@Requires(beans = { CsrfTokenRepository.class, CsrfHmacTokenGenerator.class})
+@Singleton
+public class RepositoryCsrfTokenValidator<T> implements CsrfTokenValidator<T> {
```

Comment:
Public and non-final by design?

Reply:
I made it package private and final https://github.com/micronaut-projects/micronaut-security/pull/1826/commits/816466c0eaa566f1078f6682922b2cb984aaab0f